### PR TITLE
feat: script to generate report

### PR DESCRIPTION
**Pull request by kaustavb12**
## Description

Script to generate report about the prevelance of cross-sectional report

## Testing instructions

Check the [README](https://github.com/open-craft/cross-section-link-report/blob/kaustav/generate_report_script/README.md)

## Other information

OpenCraft internal ticket : BB-7352

### CODE REVIEW

**Comment by Kelketek** [comment]
@kaustavb12 This is working for me. Please add in the counters for the other items Piotr mentioned. The only concern I have right now is if the authentication will work on the client's production instance since they have SSO, but I'm pretty sure it will, since I think the kind of authentication you're doing bypasses that. However I don't have an account to verify that with.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,16 @@
 # cross-section-link-report
 Script to generate report about the cross sectional links used in course
+
+1. Create a new file called `config.yml` based on `confiy.yml.example`. Populate all the variables there.
```
Comment (Agrendalath) [inline README.md:4]:
```suggestion
1. Create a new file called `config.yml` based on `config.yml.example`. Populate all the variables there.
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,16 @@
...
+2. Install requirements
+```
+pip install -r requirement.txt
```
Comment (Agrendalath) [inline README.md:7]: Nit: let's rename `requirement.txt` to `requirements.txt`.

File: config.yml.example [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Instance or client name to be used a report header
+INSTANCE_NAME: Test Sandbox
+REPORT_PATH: /home/reports/output
```
Comment (Agrendalath) [inline config.yml.example:3]: We could just hardcode this to `/tmp/cross-section-link-report.html`.

File: generate_report.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from helpers import login
+from helpers.helper import (
+    fetch_all_active_courses,
+    fetch_all_course_blocks,
+    fetch_block_data,
+    generate_report
+)
+
+# Login the user and fetch the session cookies
+cookie_jar = login.login_to_lms()
```
Comment (Agrendalath) [inline generate_report.py:10]:
```suggestion
session = login.login_to_lms()
```

Nit: it seems to be more self-explanatory.

File: generate_report.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+active_courses = fetch_all_active_courses(cookie_jar)
+
+course_count = 1
+for course in active_courses:
```
Comment (Agrendalath) [inline generate_report.py:15]:
```suggestion
for course_number, course in enumerate(active_courses, start=1):
```

File: generate_report.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    print(f"Course {course_count}/{len(active_courses)} : Fetching all blocks of course {course}")
```
Comment (Agrendalath) [inline generate_report.py:16]:
```suggestion
    print(f"Course {course_count}/{len(active_courses)} : Fetching all blocks for course {course}")
```

File: generate_report.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    # Fetch xblock data
+    print(f"Course {course_count}/{len(active_courses)} : Fetching xblock data for course {course.id}")
+    fetch_block_data(cookie_jar, course)
+
+    # Find and list internal cross section links
+    course.find_links_in_blocks()
+
+    course_count += 1
+
+generate_report(active_courses)
```
Comment (Agrendalath) [inline generate_report.py:31]:
```suggestion
if __name__ == "__main__":
    generate_report(active_courses)
```

Nit: we don't want to run this if this file is imported (for any reason).

File: generate_report.py (hunk 6 above)
Comment (Agrendalath) [inline generate_report.py:16]: Could we use `log.info`? It gives additional helpful information (e.g., the timestamp and code location).

File: helpers/helper.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+def fetch_all_active_courses(cookie_jar):
+    '''
+    Fetch list of all courses.
+    Discards courses whose end date is defined and is in the past.
+    Returns the other courses as active courses.
+    '''
+    active_courses = []
+
+    print("Fetching list of all courses")
+    reqs = [grequests.get(FETCH_ALL_COURSES_URL, cookies=cookie_jar)]
```
Comment (Agrendalath) [inline helpers/helper.py:43]: This API retrieves only some courses for me in the Stage environment (as a superuser). I expanded the `FETCH_ALL_COURSES_URL` URL with `?page_size=100` ([max value](https://github.com/openedx/edx-platform/blob/6adf45df2f25a0e44d08047b6b36997cab51b99c/lms/djangoapps/course_api/views.py#L317)) and got 16 results from this method.

## Comparison
### This script
```python
In [1]: from helpers.login import login_to_lms
In [2]: from helpers.helper import fetch_all_active_courses
In [3]: session = login_to_lms()
Fetching session cookies for login
Logging in using staff credentials
In [4]: courses = fetch_all_active_courses(session)
Fetching list of all courses
In [5]: len(courses)
Out[5]: 16
```

### Django shell
```python
In [1]: import pytz
In [2]: from datetime import datetime
In [3]: from django.db.models import Q
In [4]: from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
In [5]: CourseOverview.objects.filter(Q(end=None) | Q(end__gt=datetime.now(tz=pytz.UTC))).count()
Out[5]: 114
```

# Update

I modified the current script to something like this (see comments in the code):
```python
# This is important - if the `username` is not specified, the user is treated as an anonymous one.
FETCH_ALL_COURSES_URL = f"{LMS_URL}/api/courses/v1/courses/?page_size=100&username=Piotr"

def fetch_all_active_courses(cookie_jar):
    '''
    Fetch list of all courses.
    Discards courses whose end date is defined and is in the past.
    Returns the other courses as active courses.
    '''
    active_courses = []

    # This API is paginated.
    next_page = FETCH_ALL_COURSES_URL
    while next_page:
        print("Fetching list of all courses")
        reqs = [grequests.get(next_page, cookies=cookie_jar)]
        r = grequests.map(reqs)[0]

        courses = r.json().get('results')
        today = datetime.now()
        for course in courses:
            course_active = False
            end_date_str = course.get('end', None)
            if end_date_str:
                end_date_str = end_date_str.split('T')[0]
                end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
                if end_date >= today:
                    course_active = True
            else:
                course_active = True

            if course_active:
                course = Course(course)
                course.url = f"{STUDIO_URL}/course/{course.id}"
                active_courses.append(course)
        next_page = r.json().get('pagination').get('next')

    return active_courses
```

This returns 112 courses. I compared the outputs, and the discrepancy comes from the Sites (this environment has a second Site with courses filtered by an organization). In this case, it's safe to ignore.

File: helpers/helper.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    for course in courses:
+        course_active = False
+        end_date_str = course.get('end', None)
+        if end_date_str:
+            end_date_str = end_date_str.split('T')[0]
+            end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
+            if end_date >= today:
+                course_active = True
+        else:
+            course_active = True
```
Comment (Agrendalath) [inline helpers/helper.py:57]: Life's too short to parse dates manually. Just install [python-dateutil](https://pypi.org/project/python-dateutil/), import `from dateutil.parser import parse`, and do something like:
```suggestion
        try:
            course_active = parse(end_date, ignoretz=True) >= today:
        except TypeError:
            # End date is undefined.
            course_active = True
```

File: helpers/helper.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    reqs = []
+    block_count = 1
+    for usage_key in course.blocks:
+        block = course.blocks[usage_key]
+        block_data_url = f"{STUDIO_URL}/xblock/{block.usage_key}"
+        headers = {"Accept": "application/json, text/plain, */*"}
+        reqs.append(grequests.get(block_data_url, cookies=cookie_jar, headers=headers, hooks={'response': _handle_block_data(block)}))
+        block_count += 1
+    # Requests sent in batches to prevent rate limiting
+    grequests.map(reqs, size=REQUEST_BATCH_SIZE)
```
Comment (Agrendalath) [inline helpers/helper.py:76]: Does the [component_handler](https://github.com/openedx/edx-platform/blob/2f9c61e29a941545bd0216a76812c1270c5e871e/cms/djangoapps/contentstore/views/component.py#L491-L535) have rate limiting?

File: helpers/classes.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        self.id = block_dict.get('block_id', '')
+        self.lms_url = block_dict.get('lms_web_url', '')
+        self.studio_url = None
+        self.type = block_dict.get('type', '')
+        self.display_name = block_dict.get('display_name', '')
+        self.children = block_dict.get('children', None)
+        self.parent = None
+        self.data = None
+
+    def __str__(self) -> str:
```
Comment (Agrendalath) [inline helpers/classes.py:19]:
```suggestion
    def __repr__(self) -> str:
```

This will also work for debugging.

File: helpers/classes.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        self.id = course_dict['id']
+        self.url = ''
+        self.blocks_url = course_dict['blocks_url']
+        self.blocks = {}
+        self.block_id_map = {}
+        self.jump_pairs = []
+        self.other_courses = []
+        self.external_links = []
+
+    def __str__(self) -> str:
```
Comment (Agrendalath) [inline helpers/classes.py:79]:
```suggestion
    def __repr__(self) -> str:
```

This will also work for debugging.

**Review by Agrendalath — CHANGES_REQUESTED** [review]
@kaustavb12, I started testing this on Stage, and the first course (with 925 blocks) took 6 minutes to process. The CPU usage of the remote instance was ~50%. If other courses have similar sizes, it will take more than 11h to process this (assuming that we wouldn't run out of memory or file descriptors (described in the next paragraph).

I wanted to speed things up, so I changed the `REQUEST_BATCH_SIZE` to 100. It took 4 minutes to process the first course. The CPU usage of the remote instance was ~70%. However, while retrieving the blocks for the second course, I got `OSError: [Errno 24] Too many open files`. That's odd, as I set the limit for open file descriptors to `1024`. Maybe `grequests` is not closing the connections properly?

In any case, the API views are too inefficient for such a heavy operation. We may want to get the data from the `modulestore` directly instead. It's OK to run this script directly on Stage - it does not need to be a full Django app (or even a Python package) - see the scripts I linked in the Jira ticket for an example of how this could look like.

cc: @Kelketek

**Pull request closed without merging**